torch/_inductor/codegen/triton.py

```diff
@@ -281,7 +281,13 @@ def index_expr(cls, expr, dtype):
     def masked(mask, body, other):
         with V.kernel.mask_loads(mask) as new_mask:
             result = body()
-        return ops.where(new_mask, result, triton_constant(other))
+
+        # Take dtype from result to prevent accidental promotion
+        other = V.kernel.cse.generate(
+            V.kernel.compute,
+            f"tl.full({result}.shape, {triton_constant(other)}, {result}.dtype)",
+        )
```

Comment:
Simply using `triton_constant(other)` should work, as when both inputs are in the same category, no broadcasting happens.

Reply:
I'm not sure what you mean? The previous code already used `triton_constant`.

Reply:
Ah, sorry, what I meant is that it would be enough to give it the right type. It could be simply by casting the `0` to `int`, `float`, `bool` depending on the type of `result`. But we don't have a way to know the dtype of result during compilation time so I guess this is alright.

We could also always use `False`, but this would be really odd, so nvm.